### PR TITLE
[tt-crank] lower fused AdamW to ttir.adamw

### PR DESCRIPTION
**Pull request by agobeljicTT**
Lowers `aten::_fused_adamw_` (what `torch.optim.AdamW(fused=True).step()` calls in eager) to `ttir.adamw`, ttml's fused kernel. One TTIR module per call covering all params, one `ttir.adamw` each; `lr` and `step` enter as single-element tensors and `beta^t` is derived in-graph (`ttir.pow`), so the module is identical every step and the engine cache serves it after the first. Results are swapped into the caller's param/state storage. `maximize` -> `ttir.neg`; grad is cast to bf16 (kernel requirement); `grad_scale` / `found_inf` rejected.

Eager only. The `torch.compile(opt.step)` path and the in-graph input-mutation handling it needs were dropped from this PR (commit 5d75a51a7a) and go in separately: #9328 (issue) + #9326 (draft) for input mutations, #9327 (draft) for compiled AdamW, stacked on both.

Perf vs `main` (646c2d41d0), OPT_LEVEL 1, bs=1, sl=1024, 1 chip:
- Llama 3.2 1B SFT: step time -77% (3.478 s to 0.789 s), MFU 1.41% to 6.23% (+342%)
- Llama 3.1 70B: step time -57% (7.260 s to 3.153 s), MFU 0.81% to 1.87% (+131%)

Tracking: tenstorrent/tt-blacksmith#685, tenstorrent/tt-blacksmith#686.

Tests: `tests/python/op_tests/test_optimizer.py` (eager vs CPU `AdamW(fused=False)` across shapes, dtypes, options, steps, tensor lr, device placement, rejects).

### CODE REVIEW

File: tt-crank/python/tt_crank/torch/_compile.py [hunk 1]
```diff
@@ -874,6 +874,64 @@ def _(
...
+@_lowering(_aten._foreach_add.Scalar)
+@_skip_prepare(_aten._foreach_add.Scalar)
+def _(mb, tensors, scalar, alpha=1):
+    return [mb.add(t, mb.scalar_like(t, float(scalar)), float(alpha)) for t in tensors]
+
+
+def _adamw_scalar(mb, value):
+    if isinstance(value, _native.Value):
+        return value
+    return mb.full([1], float(value), _to_runtime_dtype(torch.float32))
```
[the remark this answers is not in the document]
Comment (agobeljicTT) [reply tt-crank/python/tt_crank/torch/_compile.py]: added a warning to let the user know that the tensor version is prefered

**Review by pilkicTT — CHANGES_REQUESTED** [review]
this change implements multiple unrelated things:
1. the input mutation handling + special `_copy` handling (which is suspicious)
2. the adam lowering
3. support for compiled adam

i would prefer if we would start with only 2. in this PR...

for 1., i would like us to investigate this more thouroughly and implement a more contained solution. please file an issue for this, and we will take a look - since it is core compiler behaviour, doing this within an op lowering change without sufficient tests and hollistic overview is problematic.

for 3. it is not clear to me do we really need this? and if we do, that can go in as a separate change...

File: tt-crank/src/engine/ttir_module_builder.cpp [hunk 2]
```diff
@@ -1387,6 +1396,52 @@ mlir::Value build_sdpa(ModuleBuilder &mb, mlir::Value query, mlir::Value key, ml
...
+    if (max_exp_avg_sq) {
+        result_types.push_back(max_exp_avg_sq.getType());
+        operands.push_back(max_exp_avg_sq);
+    }
+    auto attr = [&](const char *name, float value) {
+        return mb.attrs().getNamedAttr(name, mb.attrs().getF32FloatAttr(value));
+    };
+    auto op = mb.create<mlir::tt::ttir::AdamWOp>(
+        result_types, operands,
+        llvm::ArrayRef<mlir::NamedAttribute>{attr("beta1", params.beta1), attr("beta2", params.beta2),
```
Comment (acolicTT) [inline tt-crank/src/engine/ttir_module_builder.cpp]: You can create member function for this in a AdamWParams, since you have added separate struct. Also, is this named attrs required, since I don't see it for any other op?

File: tt-crank/src/torch/ops/optimizer.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        }
+    }
+
+    // Not a recompile per step: `step` and `lr` are module inputs, so the module text (and with it the
+    // engine's compilation key, a hash over the printed IR) is identical every step and the compiled
+    // program comes back from the cache. Per-step cost is building and hashing the TTIR module.
+    auto runtime_outputs = compile_and_run(std::move(mb).finalize(outputs), operands);
+    for (std::size_t k = 0; k < targets; ++k) {
+        for (std::size_t i = 0; i < n; ++i) {
+            storage_of(groups[k][i]).replace(std::move(runtime_outputs[k * n + i]));
```
Comment (acolicTT) [inline tt-crank/src/torch/ops/optimizer.cpp]: I don't like this plain replacement. Better create at::Tensors from outputs, and call replace with at::Tensor like we do in elementwise.cpp -> write_result_into, so that you have checked replace; You will have additional allocations cost, but we are not operating on that level anyway.

**Review by acolicTT — APPROVED** [review]

File: tt-crank/src/torch/ops/optimizer.cpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            build_adamw(mb, args[i], maximize ? build_neg(mb, grad) : grad, args[n + i], args[2 * n + i],
+                        amsgrad ? args[3 * n + i] : mlir::Value{}, args[(targets + 1) * n + i], args.back(), params);
+        for (std::size_t k = 0; k < targets; ++k) {
+            outputs[k * n + i] = results[k];
+        }
+    }
+
+    // Not a recompile per step: `step` and `lr` are module inputs, so the module text (and with it the
+    // engine's compilation key, a hash over the printed IR) is identical every step and the compiled
+    // program comes back from the cache. Per-step cost is building and hashing the TTIR module.
```
Comment (pilkicTT) [inline tt-crank/src/torch/ops/optimizer.cpp]: this comment only adds confusion...

File: tt-crank/src/torch/ops/optimizer.cpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    inputs.push_back(lr);
+
+    const auto operands = align_on_tt(inputs);
+    llvm::SmallVector<TensorTypeSpec, 16> specs;
+    for (const at::Tensor &t : operands) {
+        specs.push_back(spec_for(t));
+    }
+    auto mb = ModuleBuilder::init(specs);
+    const auto args = mb.args();
+    const AdamWParams params{as<float>(beta1), as<float>(beta2), as<float>(eps), as<float>(weight_decay)};
```
Comment (pilkicTT) [inline tt-crank/src/torch/ops/optimizer.cpp]: should `lr` be inside `AdamWParams`?

**Review by pilkicTT — APPROVED** [review]

File: tt-crank/src/engine/ttir_module_builder.cpp (hunk 2 above)
Comment (agobeljicTT) [reply tt-crank/src/engine/ttir_module_builder.cpp]: switched to the ods typed builder, so the named attribute list and the lambda are removed, the four f32 attrs go in directly. no need to bundle into an AdamWParams member, let me know if you really want it xd

File: tt-crank/src/torch/ops/optimizer.cpp (hunk 3 above)
Comment (agobeljicTT) [reply tt-crank/src/torch/ops/optimizer.cpp]: done. moved write_result_into out of the anonymous namespace in elementwise.cpp into tensor.hpp/tensor.cpp next to storage_of, and the optimizer now wraps each runtime output with wrap_tt_tensor and goes through it, so the swap is shape, dtype and byte count checked

File: tt-crank/src/torch/ops/optimizer.cpp (hunk 4 above)
Comment (agobeljicTT) [reply tt-crank/src/torch/ops/optimizer.cpp]: removing

File: tt-crank/src/torch/ops/optimizer.cpp (hunk 5 above)
Comment (agobeljicTT) [reply tt-crank/src/torch/ops/optimizer.cpp]: no, AdamWParams holds the four constants that become op attributes. lr is a tensor operand of ttir.adamw (a module input) so a scheduler can change it without changing the module.

Comment (agobeljicTT) [reply tt-crank/src/torch/ops/optimizer.cpp]: removed struct altogether xd

File: tt-crank/src/engine/ttir_module_builder.cpp (hunk 2 above)
Comment (agobeljicTT) [reply tt-crank/src/engine/ttir_module_builder.cpp]: removed struct xd
